### PR TITLE
Allow fastqReader and fastqWriter to work with compressed file formats.

### LINKED ISSUE #8: Support compressed fastq throughout.
Needed for upstream tools to support compressed file formats.

- [ ] Review and merge #11.
- [x] ~Review and merge https://github.com/galaxyproject/galaxy/pull/3550~
- [x] ~Review and merge https://github.com/galaxyproject/galaxy/pull/3553~
- [x] Update planemo with new XSD after #3553 so tool testing with ``decompress`` can be linted.
- [ ] Do a PyPI release with compressed IO support.
- [ ] Do a bioconda release.
- [ ] Compile a list of galaxy_sequence_utils dependent packages in tools-devteam to update (below)
- [ ] For each tool to update - add support for gz and bz2, add a test case and verify, update requirements, and delete tool_dependencies.xml.
  - [x] fastq_combiner - uses qual files and fasta files so doesn't need to be updated.
  - [ ] fastq_manipulation.xml https://github.com/galaxyproject/tools-devteam/pull/449
  - [ ] fastq_groomer
  - [ ] ...

### PR DESCRIPTION
**Pull request by jmchilton**
Fixes #8.

### CODE REVIEW

File: galaxy_utils/sequence/fastq.py [hunk 1]
```diff
@@ -535,7 +539,20 @@ def _get_med_pos( size ):
 
 
 class fastqReader( Iterator ):
-    def __init__( self, fh, format='sanger', apply_galaxy_conventions=False ):
+    def __init__( self, fh=None, format='sanger', apply_galaxy_conventions=False, path=None ):
+        if fh is None:
```
Comment (mvdbeek) [inline galaxy_utils/sequence/fastq.py]: minor detail, why not reduce the copy'n paste by making a new `get_filehandle` function with something like `self.fh = get_filehandle(fh, path))` ?

File: galaxy_utils/sequence/fastq.py [hunk 2]
```diff
@@ -696,7 +713,20 @@ def has_data( self ):
...
+            assert path is not None
+            if format.endswith(".gz"):
+                fh = gzip.GzipFile(path, "wb")
+            elif format.endswith(".bz2"):
+                fh = bz2.BZ2File(path, "wb")
+            else:
+                fh = open(path, "wb")
+        else:
+            if format.endswith(".gz"):
+                fh = gzip.GzipFile(fileobj=fh)
```
Comment (mvdbeek) [inline galaxy_utils/sequence/fastq.py]: You need to open this in `w` mode.

**Comment by peterjc** [comment]
I suspect this has not been tested on Python 3? There are annoyances with things like having to explicitly request text mode from ``gzip.open`` which defaults to binary (bytes): http://bugs.python.org/issue13989

**Comment by mvdbeek** [comment]
@jmchilton is using the `gzip.GzipFile` module, and passing along 'r' in read mode.
I think the aim is python3, if that doesn't work we'll fix it.

**Comment by peterjc** [comment]
Same issue with ``gzip.GzipFile``, it too defaults to ``rb`` and treats ``r`` as read binary https://docs.python.org/3/library/gzip.html - So I expect this to break under Python 3.

I have no objection to merging this now, and addressing Python 3 support later.

**Comment by mvdbeek** [comment]
I don't think there is a problem:
```
mvandenb@u934-bdd-4-89-6580:~/src/tag_reads/galaxy$ echo "test" > test.txt && gzip test.txt && python3 -c "import gzip;print(gzip.GzipFile('test.txt.gz').readlines())"
[b'test\n']
```

**Comment by mvdbeek** [comment]
Probably a good idea to wrap this in a TextIOWrapper in python3. That works for gzip and bzip2.

**Comment by peterjc** [comment]
Python 3's ``gzip`` does the ``TextIOWrapper`` work for you as long as ``mode="rt"`` or ``mode="rt"``, this was fixed via http://bugs.python.org/issue13989

**Comment by nsoranzo** [comment]
> I suspect this has not been tested on Python 3? There are annoyances with things like having to explicitly request text mode from `gzip.open` which defaults to binary (bytes): http://bugs.python.org/issue13989

http://bugs.python.org/issue13989 has been fixed in Python 3.2, we aim to support Python>=3.3 .

There is actually no difference for gzip.open/GzipFile  methods between Python 2 and 3, they return binary data by default (`str` in Python2, `bytes in Python3).

**Comment by peterjc** [comment]
@nsoranzo Depends on how you look at it ;)

**Comment by jmchilton** [comment]
Oh boy this is sort spiraling - in order to test the Python 3 stuff I started working on #9 and bringing the wrapper for fastq_to_tabular into the project so I can use it for testing. Getting the tests to pass required a few different changes to the rest of the code. Also here is the corresponding tool change I think https://github.com/galaxyproject/tools-devteam/pull/449/commits/a8c589d45059d6bc4a8b36bccc4b6f008623bcdd so that the script is now found on the PATH instead of being a wrapper next to the tool.

This is the "right" way to do things - bring in all the scripts, lint them all, test them all, update all the wrappers this way. It will allow a lot more reuse of these scripts and increase their quality. But is it worth all this effort - are these tools still used and useful and is Python 3 worth the time? (Unlike the core code - these could depend on Python 2 indefinitely I think).

If we think it is worth continuing to work on #9 and Python 3 - I'll bring in another tool that deals with the output stuff and test with Python 3.

**Comment by mvdbeek** [comment]
>But is it worth all this effort - are these tools still used and useful and is Python 3 worth the time?

The time spend on this now will probably be time saved in the future,
and removing the sequence_utils from galaxy is a very good step forward,
same for the compressed fastq comparison fix and the xsd update. 
For the tools themselves ... I'm not sure. I do use fastq_to_tabular occasionally
and fastq_to_fasta, but the last one can also be done with seqtk.

> If we think it is worth continuing to work on #9 and Python 3 - I'll bring in another tool that deals with the output stuff and test with Python 3.

I guess we could also add a python2 requirement and do the python3 support another time.

**Comment by peterjc** [comment]
If you can get this library to work under Python 3 (and 2), I think that'd be great.

**Comment by jmchilton** [comment]
Making more progress - I have migrated a few more tools and done some more work on Python 3 support here - with corresponding tools-devteam changes at https://github.com/galaxyproject/tools-devteam/pull/449. I'm going to step away from this until at least Monday - if someone wants a weekend project to continue working through these tools applying this pattern ;).

**Pull request merged**